src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,147 @@
+    /// presented with a progress bar. An optional <paramref name="description"/> can also be provided to update the UI
+    /// accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of items that were completed. Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItems(int count, string? description = null)
+        => new()
+        {
+            CompleteItemValue = count >= 1 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// When passed to an appropriate <see cref="IProgress{T}"/>, will indicate that all progress should be reset for
+    /// the current operation. This is normally done when the code action is performing some new phase and wishes for
+    /// the UI progress bar to restart from the beginning.
+    /// </summary>
+    /// <remarks>
+    /// Currently internal as only roslyn needs this in the impl of our suggested action (we use a progress bar to
+    /// compute the work, then reset the progress to apply all the changes).  Could be exposed later to 3rd party code
+    /// if a demonstrable need is presented.
+    /// </remarks>
+    internal static CodeAnalysisProgress Clear()
+        => new() { ClearValue = true };
+}
+
+internal sealed class CodeAnalysisProgressTracker(Action<string?, int, int>? updateAction) : IProgress<CodeAnalysisProgress>
+{
+    private string? _description;
+    private int _completedItems;
+    private int _totalItems;
+
+    public CodeAnalysisProgressTracker() : this(updateAction: null)
+    {
+    }
+
+    public string? Description
+    {
+        get => _description;
+        set
+        {
+            _description = value;
+            Update();
+        }
+    }
+
+    public int CompletedItems => _completedItems;
+
+    public int TotalItems => _totalItems;
+
+    public void AddItems(int count)
+    {
+        Interlocked.Add(ref _totalItems, count);
+        Update();
+    }
+
+    public void CompleteItems(int count)
+    {
+        Interlocked.Add(ref _completedItems, count);
+        Update();
+    }
+
+    public void Clear()
+    {
+        _totalItems = 0;
+        _completedItems = 0;
+        _description = null;
+        Update();
+    }
+
+    private void Update()
+        => updateAction?.Invoke(_description, _completedItems, _totalItems);
+
+    public void Report(CodeAnalysisProgress value)
+    {
+        if (value.DescriptionValue != null)
+            this.Description = value.DescriptionValue;
+
+        if (value.IncompleteItemsValue != null)
+            this.AddItems(value.IncompleteItemsValue.Value);
```

Comment:
>this.AddItems(value.IncompleteItemsValue.Value); [](http://example.com/codeflow?start=12&length=48)

This will end up potentially calling the Update method multiple times, should this be batched so the updateAction to prevent the updateAction getting invoked multiple times?

Reply:
This was all far too complex to begin with.  Massively simplified :)